ucp/_libs/core.pyx

```diff
@@ -0,0 +1,407 @@
+        if loop not in self.all_epoll_binded_to_event_loop:
+            loop.add_reader(self.epoll_fd, self.progress)
+            self.all_epoll_binded_to_event_loop.add(loop)
+
+    def get_ucp_worker(self):
+        return PyLong_FromVoidPtr(<void*>self.worker)
+
+
+class Endpoint:
+    """An endpoint represents a connection to a peer
+
+    Please use `create_listener()` and `create_endpoint()`
+    to create an Endpoint.
+    """
+
+    def __init__(self, ucp_endpoint, ucp_worker, send_tag,
+                 recv_tag, ctrl_send_tag, ctrl_recv_tag):
+        self._ucp_endpoint = ucp_endpoint
+        self._ucp_worker = ucp_worker
+        self._send_tag = send_tag
+        self._recv_tag = recv_tag
+        self._ctrl_send_tag = ctrl_send_tag
+        self._ctrl_recv_tag = ctrl_recv_tag
+        self._send_count = 0
+        self._recv_count = 0
+        self._closed = False
+        self.pending_msg_list = [{}]
+
+    @property
+    def uid(self):
+        """The unique ID of the endpoint"""
+        return self._recv_tag
+
+    async def signal_shutdown(self):
+        """Signal the connected peer to shutdown.
+
+        Notice, this functions doesn't close the endpoint.
+        To do that, use `.close()` or del the object.
+        """
+        if self._closed:
+            raise UCXCloseError("signal_shutdown() - Endpoint closed")
+
+        # Send a shutdown message to the peer
+        msg = np.array([42], dtype=np.uint64)
+        log = "[UCX Comm] %s ==Shutdown=> %s" % (hex(self._recv_tag),
+                                                 hex(self._send_tag))
+        logging.debug(log)
+        self.pending_msg_list.append({'log': log})
+        await tag_send(
+            self._ucp_endpoint,
+            msg, msg.nbytes,
+            self._ctrl_send_tag,
+            pending_msg=self.pending_msg_list[-1]
+        )
+
+    def closed(self):
+        """Is this endpoint closed?"""
+        return self._closed
+
+    def close(self):
+        """Close this endpoint.
+
+        Notice, this functions doesn't signal the connected peer to shutdown
+        To do that, use `.signal_shutdown()` or del the object.
+        """
+        if self._closed:
+            raise UCXCloseError("close() - Endpoint closed")
+        self._closed = True
+        logging.debug("Endpoint.close(): %s" % hex(self.uid))
+
+        for msg in self.pending_msg_list:
+            if 'future' in msg and not msg['future'].done():
+                # TODO: make sure that a potential shutdown message isn't cancelled
+                logging.debug("Future cancelling: %s" % msg['log'])
+                ucp_request_cancel(
+                    <ucp_worker_h>PyLong_AsVoidPtr(self._ucp_worker),
+                    PyLong_AsVoidPtr(msg['ucp_request'])
+                )
+
+        cdef ucp_ep_h ep = <ucp_ep_h> PyLong_AsVoidPtr(self._ucp_endpoint)
+        cdef ucs_status_ptr_t status = ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FORCE)
```

Comment:
What happens if ucp_ep_close_nb returns a ucp_request?

Reply:
The request handle leaks  :)
Fixed, thanks